crud/common/sharding_key.lua

```diff
@@ -134,12 +135,83 @@ function sharding_key_module.fetch_on_router(space_name)
     return cache.sharding_key_as_index_obj_map[space_name]
 end
 
+-- Make sure sharding key definition is a part of primary key.
+local function is_part_of_pk(space_name, primary_index_parts, sharding_key_as_index_obj)
+    dev_checks('string', 'table', 'table')
+
+    if cache.is_part_of_pk[space_name] ~= nil then
+        return cache.is_part_of_pk[space_name]
+    end
+
+    local is_part_of_pk = true
+    local pk_fieldno_map = utils.get_index_fieldno_map(primary_index_parts)
+    for _, part in ipairs(sharding_key_as_index_obj.parts) do
+        if pk_fieldno_map[part.fieldno] == nil then
+            is_part_of_pk = false
+            break
+        end
+    end
+    cache.is_part_of_pk[space_name] = is_part_of_pk
+
+    return is_part_of_pk
+end
+
+-- Build an array with sharding key values. Function extracts those values from
+-- primary key that are part of sharding key (passed as index object).
+local function extract_from_index(primary_key, primary_index_parts, sharding_key_as_index_obj)
+    dev_checks('table', 'table', 'table')
+
+    local primary_index_fieldno_map = utils.get_index_fieldno_map(primary_index_parts)
```

Comment:
We re-calculate this map on each request. At least, please, mark it as possible speed up headroom: we can cache this value.

As an option, we can do this right in the shadring key definitions fetching code and get rid of the `primary_index_parts` argument in `sharding_key_module.extract_from_pk()`.

Reply:
Added a comment. I think we can postpone it and fix later.

Reply:
Submitted a new ticket https://github.com/tarantool/crud/issues/243